x-pack/solutions/search/plugins/enterprise_search/server/deprecations/pre_eight_index_deprecator.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ElasticsearchClient } from '@kbn/core/server';
+
+const ENT_SEARCH_INDEX_PREFIX = '.ent-search-';
+
+export interface EnterpriseSearchIndexMapping {
+  name: string;
+  isDatastream: boolean;
+  datastreamName: string;
+}
+
+export async function getPreEightEnterpriseSearchIndices(
+  esClient: ElasticsearchClient
+): Promise<EnterpriseSearchIndexMapping[]> {
+  const entSearchIndices = await esClient.indices.get({
+    index: `${ENT_SEARCH_INDEX_PREFIX}*`,
+    ignore_unavailable: true,
+    expand_wildcards: ['all'],
+  });
+
+  if (!entSearchIndices) {
+    return [];
+  }
+
+  const returnIndices: EnterpriseSearchIndexMapping[] = [];
+  for (const [index, indexData] of Object.entries(entSearchIndices)) {
+    if (
+      indexData.settings?.index?.version?.created?.startsWith('7') &&
+      indexData.settings?.index?.blocks?.write !== 'true'
+    ) {
+      const dataStreamName = indexData.data_stream;
+      returnIndices.push({
+        name: index,
+        isDatastream: dataStreamName ? true : false,
+        datastreamName: dataStreamName ?? '',
+      });
+    }
+  }
+
+  return returnIndices;
+}
+
+export async function setPreEightEnterpriseSearchIndicesReadOnly(
+  esClient: ElasticsearchClient
+): Promise<string> {
+  // get the indices again to ensure nothing's changed since the last check
+  let indices = await getPreEightEnterpriseSearchIndices(esClient);
+
+  // rollover any datastreams first
+  let hasOneDatastream = false;
+  for (const index of indices) {
+    if (index.isDatastream) {
+      // let indexResponse = esClient.indices.delete({ index: index.name });
+      const indexResponse = await esClient.indices.rollover({ alias: index.datastreamName });
```

Comment:
1 data stream may have 1+ 7.x indices. The risk/side effect of looping over indices like this and calling rollover is that you might call rollover on the same data stream multiple times (thus unnecessarily creating new write indices). We may already be doing that depending on the version in which the current write index was created (maybe it was created in 8.x?), but I think it's OK to call rollover once to keep this flow simpler.

However, to avoid calling rollover multiple times on the same data stream we should keep track of all data streams that were rolled over (in a map?) and ensure you only call rollover once.